polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp

```diff
@@ -75,93 +76,297 @@ void Uniformity::print(raw_ostream &os) const { os << *this; }
+  LLVM_DEBUG(llvm::dbgs().indent(2) << "Analyzing memory effects\n");
   SmallVector<MemoryEffects::EffectInstance> effects;
   memoryEffectOp.getEffects(effects);
-  for (const auto &effect : effects) {
-    Value val = effect.getValue();
+
+  for (const auto &EI : effects) {
+    Value val = EI.getValue();
     if (!val) {
       // Memory effect on anything other than a value: conservatively assume
       // the result(s) uniformity is unknown.
       LLVM_DEBUG(llvm::dbgs() << "Memory Effect on non-values found\n");
-      for (UniformityLattice *result : results)
-        propagateIfChanged(result, result->join(Uniformity::getUnknown()));
+      return propagateAllIfChanged(results, Uniformity::getUnknown());
+    }
+
+    if (!isa<MemoryEffects::Read>(EI.getEffect()))
+      continue;
+
+    // Get the reaching def. and potential reaching def. of the value 'val' and
+    // analyze them to determine its uniformity.
+    const ReachingDefinition *rdef =
+        internalSolver.lookupState<ReachingDefinition>(op);
+    if (!rdef) {
+      LLVM_DEBUG(llvm::dbgs() << "Unable to find reaching definition\n");
+      return propagateAllIfChanged(op->getResults(), Uniformity::getUnknown());
+    }
+
+    using ModifiersTy = ReachingDefinition::ModifiersTy;
+    auto merge = [](std::optional<ModifiersTy> mods,
+                    std::optional<ModifiersTy> pMods) {
+      if (!mods && !pMods)
+        return mods;
+      if (mods && !pMods)
+        return mods;
+      if (pMods && !mods)
+        return pMods;
+      mods->merge(*pMods);
+      return mods;
+    };
+
+    // Merge mods and pMods together.
+    std::optional<ModifiersTy> mods =
+        merge(rdef->getModifiers(val), rdef->getPotentialModifiers(val));
+    if (!mods)
+      continue;
+
+    LLVM_DEBUG({
+      llvm::dbgs().indent(2) << "val: " << val << "\n";
+      llvm::dbgs().indent(2) << "merged mods:\n";
+      for (Definition mod : *mods)
+        llvm::dbgs().indent(4) << mod << "\n";
+    });
+
+    // If any of mods/pMods are dominated by a branch with a condition that is
+    // unknown/non-uniform the loaded value has the same uniformity.
+    SetVector<Value> branchConditions = collectBranchConditions(*mods);
+    if (anyOfUniformityIs(branchConditions.getArrayRef(),
+                          Uniformity::Kind::Unknown)) {
+      LLVM_DEBUG(llvm::dbgs().indent(2)
+                 << "Branch condition has unknown uniformity\n");
+      return propagateAllIfChanged(op->getResults(), Uniformity::getUnknown());
+    }
+    if (anyOfUniformityIs(branchConditions.getArrayRef(),
+                          Uniformity::Kind::NonUniform)) {
+      LLVM_DEBUG(llvm::dbgs().indent(2) << "Branch condition non-uniform\n");
+      return propagateAllIfChanged(op->getResults(),
+                                   Uniformity::getNonUniform());
+    }
+
+    // If we can't yet compute the mods/pMods operands uniformity, bail out.
+    llvm::errs() << "op: " << *op << "\n";
```

Comment:
Nit: Remove or surround with `LLVM_DEBUG` (and change to `llvm::dbgs()` in that case).